In [1]:
# Install the 'ipython-sql' and 'prettytable' libraries using pip
!pip install ipython-sql prettytable

# Import the 'prettytable' library, which is used to display data in a formatted table
import prettytable

# Set the default display format for prettytable to 'DEFAULT' (i.e., a simple table format)
prettytable.DEFAULT = 'DEFAULT'
!pip install ucimlrepo
%load_ext sql

In [2]:
import pandas as pd 
import numpy as np  
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
heart_disease = fetch_ucirepo(id=45) 
  
# data (as pandas dataframes) 
X = heart_disease.data.features 
y = heart_disease.data.targets 
  
# metadata 
print(heart_disease.metadata) 
  
# variable information 
print(heart_disease.variables) 


{'uci_id': 45, 'name': 'Heart Disease', 'repository_url': 'https://archive.ics.uci.edu/dataset/45/heart+disease', 'data_url': 'https://archive.ics.uci.edu/static/public/45/data.csv', 'abstract': '4 databases: Cleveland, Hungary, Switzerland, and the VA Long Beach', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 303, 'num_features': 13, 'feature_types': ['Categorical', 'Integer', 'Real'], 'demographics': ['Age', 'Sex'], 'target_col': ['num'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1989, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C52P4X', 'creators': ['Andras Janosi', 'William Steinbrunn', 'Matthias Pfisterer', 'Robert Detrano'], 'intro_paper': {'ID': 231, 'type': 'NATIVE', 'title': 'International application of a new probability algorithm for the diagnosis of coronary artery disease.', 'authors': 'R. Detrano, A. Jánosi, W. Steinbrunn, M

In [3]:
import csv, sqlite3

con = sqlite3.connect("SQLiteMagic.db")
cur = con.cursor()

In [4]:
%sql sqlite:///SQLiteMagic.db

In [5]:
df = pd.concat([X, y], axis=1)

In [6]:
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  num  
0  0.0   6.0    0  
1  3.0   3.0    2  
2  2.0   7.0    1  
3  0.0   3.0    0  
4  0.0   3.0    0

In [7]:
df.to_sql('heart_disease', con, if_exists='replace', index=False)

303

In [8]:
%sql SELECT * FROM heart_disease;

 * sqlite:///SQLiteMagic.db
Done.


[(63, 1, 1, 145, 233, 1, 2, 150, 0, 2.3, 3, 0.0, 6.0, 0),
 (67, 1, 4, 160, 286, 0, 2, 108, 1, 1.5, 2, 3.0, 3.0, 2),
 (67, 1, 4, 120, 229, 0, 2, 129, 1, 2.6, 2, 2.0, 7.0, 1),
 (37, 1, 3, 130, 250, 0, 0, 187, 0, 3.5, 3, 0.0, 3.0, 0),
 (41, 0, 2, 130, 204, 0, 2, 172, 0, 1.4, 1, 0.0, 3.0, 0),
 (56, 1, 2, 120, 236, 0, 0, 178, 0, 0.8, 1, 0.0, 3.0, 0),
 (62, 0, 4, 140, 268, 0, 2, 160, 0, 3.6, 3, 2.0, 3.0, 3),
 (57, 0, 4, 120, 354, 0, 0, 163, 1, 0.6, 1, 0.0, 3.0, 0),
 (63, 1, 4, 130, 254, 0, 2, 147, 0, 1.4, 2, 1.0, 7.0, 2),
 (53, 1, 4, 140, 203, 1, 2, 155, 1, 3.1, 3, 0.0, 7.0, 1),
 (57, 1, 4, 140, 192, 0, 0, 148, 0, 0.4, 2, 0.0, 6.0, 0),
 (56, 0, 2, 140, 294, 0, 2, 153, 0, 1.3, 2, 0.0, 3.0, 0),
 (56, 1, 3, 130, 256, 1, 2, 142, 1, 0.6, 2, 1.0, 6.0, 2),
 (44, 1, 2, 120, 263, 0, 0, 173, 0, 0.0, 1, 0.0, 7.0, 0),
 (52, 1, 3, 172, 199, 1, 0, 162, 0, 0.5, 1, 0.0, 7.0, 0),
 (57, 1, 3, 150, 168, 0, 0, 174, 0, 1.6, 1, 0.0, 3.0, 0),
 (48, 1, 2, 110, 229, 0, 0, 168, 0, 1.0, 3, 0.0, 7.0, 1),
 (54, 1, 4, 140, 239, 0, 0, 160, 0, 1.2, 1, 0.0, 3.0, 0),
 (48, 0, 3, 130, 275, 0, 0, 139, 0, 0.2, 1, 0.0, 3.0, 0),
 (49, 1, 2, 130, 266, 0, 0, 171, 0, 0.6, 1, 0.0, 3.0, 0),
 (64, 1, 1, 110, 211, 0, 2, 144, 1, 1.8, 2, 0.0, 3.0, 0),
 (58, 0, 1, 150, 283, 1, 2, 162, 0, 1.0, 1, 0.0, 3.0, 0),
 (58, 1, 2, 120, 284, 0, 2, 160, 0, 1.8, 2, 0.0, 3.0, 1),
 (58, 1, 3, 132, 224, 0, 2, 173, 0, 3.2, 1, 2.0, 7.0, 3),
 (60, 1, 4, 130, 206, 0, 2, 132, 1, 2.4, 2, 2.0, 7.0, 4),
 (50, 0, 3, 120, 219, 0, 0, 158, 0, 1.6, 2, 0.0, 3.0, 0),
 (58, 0, 3, 120, 340, 0, 0, 172, 0, 0.0, 1, 0.0, 3.0, 0),
 (66, 0, 1, 150, 226, 0, 0, 114, 0, 2.6, 3, 0.0, 3.0, 0),
 (43, 1, 4, 150, 247, 0, 0, 171, 0, 1.5, 1, 0.0, 3.0, 0),
 (40, 1, 4, 110, 167, 0, 2, 114, 1, 2.0, 2, 0.0, 7.0, 3),
 (69, 0, 1, 140, 239, 0, 0, 151, 0, 1.8, 1, 2.0, 3.0, 0),
 (60, 1, 4, 117, 230, 1, 0, 160, 1, 1.4, 1, 2.0, 7.0, 2),
 (64, 1, 3, 140, 335, 0, 0, 158, 0, 0.0, 1, 0.0, 3.0, 1),
 (59, 1, 4, 135, 234, 0, 0, 161, 0, 0.5, 2, 0.0, 7.0, 0),
 (44, 1, 3, 130, 233, 0, 0, 179, 1, 0.4, 1, 0.0, 3.0, 0),
 (42, 1, 4, 140, 226, 0, 0, 178, 0, 0.0, 1, 0.0, 3.0, 0),
 (43, 1, 4, 120, 177, 0, 2, 120, 1, 2.5, 2, 0.0, 7.0, 3),
 (57, 1, 4, 150, 276, 0, 2, 112, 1, 0.6, 2, 1.0, 6.0, 1),
 (55, 1, 4, 132, 353, 0, 0, 132, 1, 1.2, 2, 1.0, 7.0, 3),
 (61, 1, 3, 150, 243, 1, 0, 137, 1, 1.0, 2, 0.0, 3.0, 0),
 (65, 0, 4, 150, 225, 0, 2, 114, 0, 1.0, 2, 3.0, 7.0, 4),
 (40, 1, 1, 140, 199, 0, 0, 178, 1, 1.4, 1, 0.0, 7.0, 0),
 (71, 0, 2, 160, 302, 0, 0, 162, 0, 0.4, 1, 2.0, 3.0, 0),
 (59, 1, 3, 150, 212, 1, 0, 157, 0, 1.6, 1, 0.0, 3.0, 0),
 (61, 0, 4, 130, 330, 0, 2, 169, 0, 0.0, 1, 0.0, 3.0, 1),
 (58, 1, 3, 112, 230, 0, 2, 165, 0, 2.5, 2, 1.0, 7.0, 4),
 (51, 1, 3, 110, 175, 0, 0, 123, 0, 0.6, 1, 0.0, 3.0, 0),
 (50, 1, 4, 150, 243, 0, 2, 128, 0, 2.6, 2, 0.0, 7.0, 4),
 (65, 0, 3, 140, 417, 1, 2, 157, 0, 0.8, 1, 1.0, 3.0, 0),
 (53, 1, 3, 130, 197, 1, 2, 152, 0, 1.2, 3, 0.0, 3.0, 0),
 (41, 0, 2, 105, 198, 0, 0, 168, 0, 0.0, 1, 1.0, 3.0, 0),
 (65, 1, 4, 120, 177, 0, 0, 140, 0, 0.4, 1, 0.0, 7.0, 0),
 (44, 1, 4, 112, 290, 0, 2, 153, 0, 0.0, 1, 1.0, 3.0, 2),
 (44, 1, 2, 130, 219, 0, 2, 188, 0, 0.0, 1, 0.0, 3.0, 0),
 (60, 1, 4, 130, 253, 0, 0, 144, 1, 1.4, 1, 1.0, 7.0, 1),
 (54, 1, 4, 124, 266, 0, 2, 109, 1, 2.2, 2, 1.0, 7.0, 1),
 (50, 1, 3, 140, 233, 0, 0, 163, 0, 0.6, 2, 1.0, 7.0, 1),
 (41, 1, 4, 110, 172, 0, 2, 158, 0, 0.0, 1, 0.0, 7.0, 1),
 (54, 1, 3, 125, 273, 0, 2, 152, 0, 0.5, 3, 1.0, 3.0, 0),
 (51, 1, 1, 125, 213, 0, 2, 125, 1, 1.4, 1, 1.0, 3.0, 0),
 (51, 0, 4, 130, 305, 0, 0, 142, 1, 1.2, 2, 0.0, 7.0, 2),
 (46, 0, 3, 142, 177, 0, 2, 160, 1, 1.4, 3, 0.0, 3.0, 0),
 (58, 1, 4, 128, 216, 0, 2, 131, 1, 2.2, 2, 3.0, 7.0, 1),
 (54, 0, 3, 135, 304, 1, 0, 170, 0, 0.0, 1, 0.0, 3.0, 0),
 (54, 1, 4, 120, 188, 0, 0, 113, 0, 1.4, 2, 1.0, 7.0, 2),
 (60, 1, 4, 145, 282, 0, 2, 142, 1, 2.8, 2, 2.0, 7.0, 2),
 (60, 1, 3, 140, 185, 0, 2, 155, 0, 3.0, 2, 0.0, 3.0, 1),
 (54, 1, 3, 150, 232, 0, 2, 165, 0, 1.6, 1, 0.0, 7.0, 0),
 (59, 1, 4, 170, 326, 0, 2, 140, 1, 3.4, 3, 0.0, 7.0, 2)

In [9]:
%%sql SELECT 
    MIN(age) AS min_age,
    MAX(age) AS max_age,
    ROUND(AVG(age), 0) AS avg_age
FROM 
    heart_disease; 

 * sqlite:///SQLiteMagic.db
Done.


[(29, 77, 54.0)]

In [10]:
%%sql SELECT
CASE
    WHEN sex = 0 THEN 'female'
    WHEN sex = 1 THEN 'male'
END AS gender,
 COUNT(*) AS patient_count
FROM heart_disease
GROUP BY gender;


 * sqlite:///SQLiteMagic.db
Done.


[('female', 97), ('male', 206)]

In [11]:
%%sql
SELECT
CASE
    WHEN cp = 1 THEN 'typical angina'
    WHEN cp = 2 THEN 'atypical angina'
    WHEN cp = 3 THEN 'non-anginal pain'
    WHEN cp = 4 THEN 'asymptomatic'
END
cp, COUNT(*)  AS pain_frequency
FROM heart_disease
GROUP BY cp;


 * sqlite:///SQLiteMagic.db
Done.


[('typical angina', 23),
 ('atypical angina', 50),
 ('non-anginal pain', 86),
 ('asymptomatic', 144)]

In [12]:
%%sql
SELECT
CASE
        WHEN age BETWEEN 20 AND 30 THEN '20-30'
        WHEN age BETWEEN 31 AND 40 THEN '31-40'
        WHEN age BETWEEN 41 AND 50 THEN '41-50'
        WHEN age BETWEEN 51 AND 60 THEN '51-60'
        WHEN age BETWEEN 61 AND 70 THEN '61-70'
        ELSE 'Above 70'
END AS age_group, 
SUM(CASE WHEN num = 1 THEN 1 ELSE 0 END) AS heart_disease_count,
SUM(CASE WHEN num = 0 THEN 1 ELSE 0 END) AS healthy_count
FROM heart_disease
GROUP BY age_group
ORDER BY age_group;



 * sqlite:///SQLiteMagic.db
Done.


[('20-30', 0, 1),
 ('31-40', 3, 11),
 ('41-50', 11, 53),
 ('51-60', 27, 64),
 ('61-70', 14, 30),
 ('Above 70', 0, 5)]

In [13]:
%%sql
SELECT MAX(chol) AS max_cholesterol , MIN(chol) min_cholesterol
FROM Heart_disease;

 * sqlite:///SQLiteMagic.db
Done.


[(564, 126)]

In [14]:
%%sql
SELECT
CASE
    WHEN sex = 0 THEN 'female'
    WHEN sex = 1 THEN 'male'
END AS gender,
MIN(age) AS youngest, MAX(age) AS oldest
FROM heart_disease 
GROUP BY gender;


 * sqlite:///SQLiteMagic.db
Done.


[('female', 34, 76), ('male', 29, 77)]

Determine the age range (youngest and oldest) for male and female patients separately.

Investigate the distribution of the target variable (presence or absence of heart disease) within different age groups (e.g., 20-30, 30-40, etc.).

In [17]:
%%sql
SELECT
CASE
    WHEN age BETWEEN 20 AND 30 THEN '20-30'
    WHEN age BETWEEN 31 AND 40 THEN '31-40'
    WHEN age BETWEEN 41 AND 50 THEN '41-50'
    WHEN age BETWEEN 51 AND 60 THEN '51-60'
    WHEN age BETWEEN 61 AND 70 THEN '61-70'
    ELSE 'Above 70'
END AS age_group, 
COUNT(num)
FROM heart_disease 
GROUP BY age_group;

 * sqlite:///SQLiteMagic.db
Done.


[('20-30', 1),
 ('31-40', 17),
 ('41-50', 76),
 ('51-60', 130),
 ('61-70', 73),
 ('Above 70', 6)]

Find the maximum heart rate achieved during exercise for different age groups (e.g., 30-40, 40-50, etc.).

In [19]:
%%sql
SELECT
CASE
    WHEN age BETWEEN 20 AND 30 THEN '20-30'
    WHEN age BETWEEN 31 AND 40 THEN '31-40'
    WHEN age BETWEEN 41 AND 50 THEN '41-50'
    WHEN age BETWEEN 51 AND 60 THEN '51-60'
    WHEN age BETWEEN 61 AND 70 THEN '61-70'
    ELSE 'Above 70'
END AS age_group,
MAX(thalach) AS MAX_heart_rate 
FROM heart_disease 
GROUP BY age_group;

 * sqlite:///SQLiteMagic.db
Done.


[('20-30', 202),
 ('31-40', 192),
 ('41-50', 194),
 ('51-60', 195),
 ('61-70', 179),
 ('Above 70', 162)]

Calculate the percentage of patients with fasting blood sugar greater than 120 mg/dl.

In [21]:
%%sql
SELECT COUNT(*) AS total_patients,
ROUND(CAST((SELECT COUNT(fbs) FROM heart_disease WHERE fbs = 1) AS REAL)*100/COUNT(*), 2)  AS perc_diabetes_patients
FROM heart_disease 

;

 * sqlite:///SQLiteMagic.db
Done.


[(303, 14.85)]

Find the ratio of patients with abnormal resting electrocardiographic results to those with normal results.

In [69]:
%%sql 
SELECT 
DISTINCT ROUND(CAST((SELECT COUNT(restecg) FROM heart_disease WHERE restecg = 1) AS REAL) *100/ CAST((SELECT COUNT(restecg) FROM heart_disease WHERE restecg = 0) AS REAL), 1)  AS ratio_abn_ecg                                         
 FROM heart_disease;

 * sqlite:///SQLiteMagic.db
Done.


[(2.6,)]

Count the number of patients with reversible thalassemia detected by thallium stress testing.

In [82]:
%%sql 
SELECT COUNT(*) AS num_pat
FROM heart_disease 
WHERE thal = 7;

 * sqlite:///SQLiteMagic.db
Done.


[(117,)]

Calculate the average age of patients who experienced chest pain during diagnosis.

In [89]:
%%sql
SELECT ROUND(AVG(age), 0) AS avg_age_exang
FROM heart_disease 
WHERE exang = 1;

 * sqlite:///SQLiteMagic.db
Done.


[(56.0,)]

Investigate the distribution of patients based on the number of major vessels colored by fluoroscopy (0-3).

In [92]:
%%sql
SELECT ca, COUNT(*) AS num_pat
FROM heart_disease 
GROUP BY ca;

 * sqlite:///SQLiteMagic.db
Done.


[(None, 4), (0.0, 176), (1.0, 65), (2.0, 38), (3.0, 20)]